src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers/ClientMethodsAnalyzer.cs

```diff
@@ -21,104 +28,126 @@ public class ClientMethodsAnalyzer : ClientAnalyzerBase
             Descriptors.AZC0015
         });
 
-        private static void CheckClientMethod(ISymbolAnalysisContext context, IMethodSymbol member)
+        private static void CheckCancellationTokenOrRequestContext(ISymbolAnalysisContext context, IMethodSymbol member)
         {
-            static bool SupportsCancellationsParameter(IParameterSymbol parameterSymbol)
+            static bool IsRequestContext(IParameterSymbol parameterSymbol)
             {
-                return (parameterSymbol.Name == "cancellationToken" && parameterSymbol.Type.Name == "CancellationToken") ||
-                       (parameterSymbol.Name == "context" && parameterSymbol.Type.Name == "RequestContext");
+                return parameterSymbol.Name == "context" && parameterSymbol.Type.Name == "RequestContext";
             }
 
-            CheckClientMethodReturnType(context, member);
+            static bool IsCancellationToken(IParameterSymbol parameterSymbol)
+            {
+                return parameterSymbol.Name == "cancellationToken" && parameterSymbol.Type.Name == "CancellationToken";
+            }
 
-            if (!member.IsVirtual && !member.IsOverride)
+            static bool IsCancellationOrRequestContext(IParameterSymbol parameterSymbol)
             {
-                context.ReportDiagnostic(Diagnostic.Create(Descriptors.AZC0003, member.Locations.First()), member);
+                return IsCancellationToken(parameterSymbol) || IsRequestContext(parameterSymbol);
             }
 
             var lastArgument = member.Parameters.LastOrDefault();
```

Comment:
Should we check early that the member has at least 1 Parameter and bail out early if so? This would avoid a few null checks below for `lastArgument` as well as any other analysis of the parameters.

Reply:
There are no checks below if `lastArgument` is null. What additional checks you are mentioning?

Reply:
Doesn't the line right below check?

```c#
var isCancellationOrRequestContext = lastArgument != null && IsCancellationOrRequestContext(lastArgument);
```

If `members.Parameters` has no members, the rest of these checks are not necessary.

Reply:
I'm afraid we cannot. If `members.Parameters` has no members, we still want to find the overload below:
```C#
var overloadSupportsCancellations = FindMethod(
    member.ContainingType.GetMembers(member.Name).OfType<IMethodSymbol>(),
    member.TypeParameters,
    member.Parameters,
    p => IsCancellationToken(p));
```